src/lib/duckdb_utils.py

```diff
@@ -0,0 +1,167 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility functions for working with DuckDB databases."""
+
+import duckdb
+from openrelik_ai_common.providers import manager
+
+
+def is_sql_file(magic_text: str) -> bool:
+    """Check if file is a valid SQL database file based on magic text.
+
+    Args:
+        magic_text: The magic text of the file.
+
+    Returns:
+        bool: True if file is a valid SQL database file.
+    """
+    if not magic_text:
+        return False
+
+    magic_text_lower = magic_text.lower()
+
+    # Define supported SQL database types to check against the mime magic text
+    sql_database_types = ["sqlite", "duckdb"]
+
+    # Check if the magic text contains "database" and any of the supported SQL database types
+    contains_database = "database" in magic_text_lower
+    contains_supported_type = any(db_type in magic_text_lower for db_type in sql_database_types)
+
+    return contains_database and contains_supported_type
+
+
+def get_tables_schemas(file: object) -> dict:
+    """Get all tables and their schemas from the DuckDB database.
+
+    Args:
+        file: A SQL database file.
+
+    Returns:
+        dict: A dictionary where keys are table names and values are dictionaries of column names and their
+    """
+    if not is_sql_file(file.magic_text):
+        return {}
+
+    db_conn = duckdb.connect(file.path, read_only=True)
+
+    # Query to get all tables and their schemas using DuckDB's information schema.
+    tables_schemas_query = """
+        SELECT t.table_name, c.column_name, c.data_type
+        FROM information_schema.tables t
+        JOIN information_schema.columns c
+            ON t.table_schema = c.table_schema
+            AND t.table_name = c.table_name
+        WHERE t.table_type = 'BASE TABLE'
+        ORDER BY t.table_schema, t.table_name, c.ordinal_position;
+    """
+
+    try:
+        # Use .fetchall() to get all rows as a list of tuples
+        results = db_conn.execute(tables_schemas_query).fetchall()
+        tables_schemas = {}
+        for table_name, column_name, data_type in results:
+            if table_name not in tables_schemas:
+                tables_schemas[table_name] = {}
+            tables_schemas[table_name][column_name] = data_type
+    except Exception as e:
+        raise RuntimeError(e)
+    finally:
+        db_conn.close()
+
+    return tables_schemas
+
+
+def run_query(file: object, sql_query: str) -> list[dict]:
+    """Run a SQL query on a DuckDB database file and return a list of dictionaries.
+
+    Args:
+        file: A SQL database file.
+        sql_query: The SQL query to execute.
+
+    Returns:
+        list[dict]: The result of the query as a list of dictionaries.
+    """
+    if not is_sql_file(file.magic_text):
+        raise RuntimeError("File is not a supported SQL format.")
+
+    db_conn = duckdb.connect(file.path, read_only=True)
```

Comment:
Have a look at the options below but my advice would be to set (at least) `enable_external_access` to False to (among other things) disallow queries to copy data to external files (or load external files....).


https://duckdb.org/docs/stable/configuration/overview.html#configuration-reference

Reply:
done